src/datasource/gateway.js

```diff
@@ -0,0 +1,273 @@
+          page,
+          per_page,
+        }
+      }
+    }),
+  }
+
+  use(provider) {
+    this.provider = provider;
+    return this;
+  }
+
+  // looks ugly with the ifs
+  // but the gateway constantly responds with 200
+  // and responds with errors in response.data
+  // in a non-consistent form.
+  parseSuccessfulResponse = (response) => {
+    
+    if (typeof response.data === 'string' && response.data.indexOf('Method Not Allowed') > -1) {
+      throw {
+        message: 'Method Not Allowed',
+      }
+    }
+
+    if (response.data.code && response.data.code !== 200) {
+      throw response.data;
+    }
+
+    return response.data;
+  }
+
+  parseErrorResponse = (error) => {
+    if (error.response && error.response.data && error.response.data.error) {
+      throw error.response.data.error;
+    }
+
+    if (typeof error === 'string') {
+      throw {
+        message: error,
+      }
+    }
+
+    throw error;
+
+  }
+
+  perform = async (requestName, ...args) => {
+    const reqConfig = typeof this.requests[requestName] === 'function' ?
+      this.requests[requestName](...args) :
+      this.requests[requestName];
+
+    const response = await this
+      .provider
+      .http
+      .request(reqConfig)
+      .then(this.parseSuccessfulResponse)
+      .catch(this.parseErrorResponse);
+
+    return response;
+  }
+
+  // does not really need to be bound to `this`, but keeping it for semantics' sake.
+  // arguments explicit forward for clear signature lookup, avoid using `...args` 
+  _getHeight = this.perform.bind(this, 'getHeight');
+  _getAccount = (id) => this.perform.call(this, 'getAccount', id);
+  _getTransaction = (id) => this.perform.call(this, 'getTransaction', id);
+  _getBlock = (height) => this.perform.call(this, 'getBlock', height);
+  _getBlockTxs = (height, prove, page, per_page) => this.perform.call(this, 'getBlockTxs', height, prove, page, per_page);
+  _getApps = (staking_status, height, blockchain, page, per_page) => this.perform.call(this, 'getApps', staking_status, height, blockchain, page, per_page);
+  _getSupply = this.perform.bind(this, 'getSupply');
+  _getNodes = (staking_status, jailed_status, height, blockchain, page, per_page) => this.perform.call(this, 'getNodes', staking_status, jailed_status, height, blockchain, page, per_page);
+
+  // For semantic separation, and for "ease of middlewaring" when necessary.
+  query = {
+    getHeight: this._getHeight,
+    getAccount: this._getAccount,
+    getTransaction: this._getTransaction,
+    getBlock: this._getBlock,
+    getBlockTxs: this._getBlockTxs,
+    getApps: this._getApps,
+    getSupply: this._getSupply,
+    getNodes: this._getNodes,
+  }
+}
+
+/**
+ * Exposes registered/allowed gateway queries.
+ * This layer is added for gateway level logic control.
+ */
+class GatewayClient {
+  constructor(httpProvider, requestsController, config) {
+    this.controller = requestsController.use(httpProvider)
+    this.config     = config;
```

Comment:
Woah what's this huge space here haha, let's run the prettier on it!

Reply:
That was like, intentionally huge, hahaha!

Reply:
Ohh you're lining these up on purpose, Go-style?

Reply:
Yes sir, no particular thought process in here, just in anarchy in the absence of a linter. Let's define a lint style though!

Reply:
Oh I'd just go with the default prettier values for sure. We can lint in another PR though to not introduce clutter into this one.